File: admin-dev/server-python/app/services/frontmatter.py
"""
Front Matter 解析与生成服务
对齐 Node.js 后端的输出格式
"""
from __future__ import annotations

import re
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

import frontmatter


# ============================================================
# 解析
# ============================================================

def parse_frontmatter(content: str) -> Tuple[Dict[str, Any], str]:
    """
    解析 Markdown 文件的 Front Matter。

    Returns:
        (front_matter_dict, body_text)
    """
    content = content.lstrip("﻿")  # 去 BOM
    post = frontmatter.loads(content)
    return dict(post.metadata), post.content


def flatten_categories(raw: Any) -> List[str]:
    """
    将 Front Matter 中的 categories 统一为扁平列表。

    支持以下格式：
      - [技术研习, Vue3]         → ["技术研习", "Vue3"]
      - [[技术研习, Vue3]]        → ["技术研习", "Vue3"]
      - 技术研习                  → ["技术研习"]
    """
    if raw is None:
        return []
    if isinstance(raw, str):
        return [raw]
    if isinstance(raw, list):
        if len(raw) == 0:
            return []
        # 嵌套数组：[[一级, 二级]]
        if isinstance(raw[0], list):
            return [item for sub in raw for item in sub]
        # 扁平列表：[一级, 二级]
        return list(raw)
    return [str(raw)]


# ============================================================
# 生成
# ============================================================

def _escape_yaml_value(val: str) -> str:
    """如果字符串包含 YAML 特殊字符则加引号"""
    if not val:
        return val
    special = r"""[:#\n\r\t{}[\],&*?|<>=!%@`"]"""
    if re.search(special, val) or val.startswith(" ") or val.endswith(" "):
        return f'"{val}"'
    return val


def generate_frontmatter(data: Dict[str, Any]) -> str:
    """
    将 Front Matter 字典序列化为 YAML 文本（包含 --- 分隔符）。
    输出格式与 Node.js 后端的 generateFrontMatter 完全一致。
    """
    lines: List[str] = ["---"]

    # 基础字段
    if data.get("title"):
        lines.append(f"title: {_escape_yaml_value(str(data['title']))}")
    if data.get("date"):
        lines.append(f"date: {data['date']}")
    if data.get("updated"):
        lines.append(f"updated: {data['updated']}")

    # 分类 —— 使用嵌套数组格式 [[一级, 二级]]
    categories = data.get("categories")
    if categories:
        lines.append("")
        lines.append("categories:")
        if isinstance(categories, list) and len(categories) > 0:
            if isinstance(categories[0], list):
                # 已经是嵌套格式
                for cat in categories:
                    lines.append(f"  - [{', '.join(str(c) for c in cat)}]")
            else:
                # 扁平格式 → 嵌套
                lines.append(f"  - [{', '.join(str(c) for c in categories)}]")

    # 标签
    tags = data.get("tags")
    if tags and len(tags) > 0:
        lines.append("")
        lines.append("tags:")
        for tag in tags:
            lines.append(f"  - {tag}")

    # 可选字段（保持与 Node 后端一致的输出顺序）
    optional_fields = [
        ("description", "description"),
        ("layout", "layout"),
        ("comments", "comments"),
        ("permalink", "permalink"),
        ("excerpt", "excerpt"),
        ("published", "published"),
        ("lang", "lang"),
        ("cover", "cover"),
        ("sticky", "sticky"),
        ("slug", "slug"),
        ("status", "status"),
        ("series", "series"),
    ]
    for key, field_name in optional_fields:
        val = data.get(field_name if key == field_name else field_name)
        if val is not None:
            lines.append("")
            if isinstance(val, str):
                lines.append(f"{field_name}: {_escape_yaml_value(val)}")
            else:
                lines.append(f"{field_name}: {val}")

    # 系列顺序（紧跟 series）
    if data.get("series_order") is not None:
        lines.append(f"series_order: {data['series_order']}")

    lines.append("---")
    lines.append("")
    return "\n".join(lines)


# ============================================================
# 图片相关工具
# ============================================================

def get_asset_folder_path(article_path: str, posts_dir: Path) -> Path:
    """获取文章对应的资源文件夹路径（与文章同名，去掉 .md 扩展名）"""
    return posts_dir / article_path.replace(".md", "")


def get_next_image_number(article_path: str, posts_dir: Path) -> int:
    """获取下一个图片编号（从 1 开始递增）"""
    asset_folder = get_asset_folder_path(article_path, posts_dir)
    if not asset_folder.exists():
        return 1

    max_num = 0
    for f in asset_folder.iterdir():
        if f.is_file() and re.match(r".*img(\d+)\.", f.name, re.IGNORECASE):
            m = re.search(r"img(\d+)", f.name, re.IGNORECASE)
            if m:
                max_num = max(max_num, int(m.group(1)))
    return max_num + 1

File: admin-dev/frontend/src/utils/frontmatter.ts
/**
 * Front Matter 解析与序列化工具
 * 浏览器兼容，不依赖 Node.js 模块
 */

/**
 * 轻量 Front Matter 解析器（替代 gray-matter）
 * 仅支持博客用到的简单 YAML 格式
 */
function parseMatter(content: string): { data: Record<string, unknown>; content: string } {
  const trimmed = content.replace(/^﻿/, '')
  const match = trimmed.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n?/)
  if (!match) {
    return { data: {}, content: trimmed }
  }
  const yaml = match[1]
  const body = trimmed.slice(match[0].length)
  return { data: parseSimpleYaml(yaml), content: body }
}

/**
 * 简易 YAML 解析器
 * 支持 key: value、列表、嵌套数组 [[a, b]] 格式
 */
function parseSimpleYaml(yaml: string): Record<string, unknown> {
  const result: Record<string, unknown> = {}
  const lines = yaml.split(/\r?\n/)
  let currentKey = ''
  let currentList: unknown[] = []

  function flushList() {
    if (currentKey && currentList.length > 0) {
      result[currentKey] = currentList
      currentKey = ''
      currentList = []
    }
  }

  for (const rawLine of lines) {
    // 跳过空行
    if (rawLine.trim() === '') continue

    // 列表项：  - value 或  - [a, b]
    const listMatch = rawLine.match(/^(\s+)-\s+(.*)$/)
    if (listMatch) {
      let val: unknown = listMatch[2].trim()
      // 嵌套数组格式：[a, b]
      const arrMatch = (val as string).match(/^\[(.*)\]$/)
      if (arrMatch) {
        val = arrMatch[1].split(',').map(s => s.trim())
      }
      // 自动转换布尔值和数字
      else if (val === 'true') val = true
      else if (val === 'false') val = false
      else if (/^\d+$/.test(val as string)) val = Number(val)
      // 去除引号
      else {
        val = unwrapQuotes(val as string)
      }
      currentList.push(val)
      continue
    }

    // 新的 key: value 行，先 flush 之前的列表
    flushList()

    const kvMatch = rawLine.match(/^([a-zA-Z_][a-zA-Z0-9_]*)\s*:\s*(.*)$/)
    if (kvMatch) {
      const key = kvMatch[1]
      let val = kvMatch[2].trim()
      if (val === '') {
        // 空 value → 后面跟列表
        currentKey = key
        currentList = []
        continue
      }
      // 自动转换类型
      if (val === 'true') result[key] = true
      else if (val === 'false') result[key] = false
      else if (/^\d+$/.test(val)) result[key] = Number(val)
      else {
        // 内联数组格式：[a, b, c]
        const inlineArrMatch = val.match(/^\[(.*)\]$/)
        if (inlineArrMatch) {
          result[key] = inlineArrMatch[1].split(',').map(s => unwrapQuotes(s.trim()))
        } else {
          result[key] = unwrapQuotes(val)
        }
      }
    }
  }

  flushList()
  return result
}

/**
 * 去除 YAML 字符串两端的引号
 */
function unwrapQuotes(val: string): string {
  if (
    (val.startsWith('"') && val.endsWith('"')) ||
    (val.startsWith("'") && val.endsWith("'"))
  ) {
    return val.slice(1, -1)
  }
  return val
}
import type { FrontMatter } from '@/types/api'

/**
 * 解析 Markdown 文件的 Front Matter
 * @param content Markdown 文件完整内容
 * @returns 解析后的 Front Matter 和正文内容
 */
export function parseFrontMatter(content: string): {
  frontMatter: FrontMatter
  body: string
} {
  try {
    const { data, content: body } = parseMatter(content)
    return {
      frontMatter: data as unknown as FrontMatter,
      body
    }
  } catch (error) {
    console.error('解析 Front Matter 失败:', error)
    throw new Error('文章格式错误：无法解析 Front Matter')
  }
}

/**
 * 序列化 Front Matter 为 YAML 格式
 * @param frontMatter Front Matter 对象
 * @returns YAML 格式的字符串（不包含 --- 分隔符）
 */
export function serializeFrontMatter(frontMatter: Partial<FrontMatter>): string {
  const lines: string[] = []

  // 标题（必须）
  lines.push(`title: ${escapeYamlString(frontMatter.title || '未命名文章')}`)

  // 日期（必须）
  const dateStr = frontMatter.date || new Date().toISOString().replace('T', ' ').slice(0, 19)
  lines.push(`date: ${dateStr}`)

  // 更新日期（可选）
  if (frontMatter.updated) {
    lines.push(`updated: ${frontMatter.updated}`)
  }

  lines.push('')

  // 分类（必须）— 嵌套数组格式 [[一级, 二级]]
  lines.push('categories:')
  if (Array.isArray(frontMatter.categories) && frontMatter.categories.length > 0) {
    // 如果已有嵌套数组格式（categories[0] 是数组），直接使用
    if (Array.isArray(frontMatter.categories[0])) {
      for (const cat of frontMatter.categories) {
        const catArr = cat as unknown as string[]
        lines.push(`  - [${catArr.map(c => escapeYamlString(c)).join(', ')}]`)
      }
    } else {
      // 扁平列表格式，转换为嵌套数组 [[一级, 二级]]
      const cats = frontMatter.categories as string[]
      lines.push(`  - [${cats.map(c => escapeYamlString(c)).join(', ')}]`)
    }
  } else {
    lines.push('  - [未分类]')
  }

  lines.push('')

  // 标签（可选）
  lines.push('tags:')
  if (Array.isArray(frontMatter.tags) && frontMatter.tags.length > 0) {
    frontMatter.tags.forEach((tag) => {
      lines.push(`  - ${escapeYamlString(tag)}`)
    })
  }

  // 描述（可选）
  if (frontMatter.description) {
    lines.push('')
    lines.push(`description: ${escapeYamlString(frontMatter.description)}`)
  }

  // 布局（可选）
  if (frontMatter.layout) {
    lines.push('')
    lines.push(`layout: ${frontMatter.layout}`)
  }

  // 评论（可选）
  if (frontMatter.comments !== undefined) {
    lines.push('')
    lines.push(`comments: ${frontMatter.comments}`)
  }

  // 永久链接（可选）
  if (frontMatter.permalink) {
    lines.push('')
    lines.push(`permalink: ${frontMatter.permalink}`)
  }

  // 摘录（可选）
  if (frontMatter.excerpt) {
    lines.push('')
    lines.push(`excerpt: ${escapeYamlString(frontMatter.excerpt)}`)
  }

  // 发布状态（可选）
  if (frontMatter.published !== undefined) {
    lines.push('')
    lines.push(`published: ${frontMatter.published}`)
  }

  // 语言（可选）
  if (frontMatter.lang) {
    lines.push('')
    lines.push(`lang: ${frontMatter.lang}`)
  }

  // 封面图（可选）
  if (frontMatter.cover) {
    lines.push('')
    lines.push(`cover: ${frontMatter.cover}`)
  }

  // 置顶（可选）
  if (frontMatter.sticky !== undefined) {
    lines.push('')
    lines.push(`sticky: ${frontMatter.sticky}`)
  }

  // Slug（可选）
  if (frontMatter.slug) {
    lines.push('')
    lines.push(`slug: ${frontMatter.slug}`)
  }

  // 状态（可选）
  if (frontMatter.status) {
    lines.push('')
    lines.push(`status: ${frontMatter.status}`)
  }

  // 系列（可选）
  if (frontMatter.series) {
    lines.push('')
    lines.push(`series: ${escapeYamlString(frontMatter.series)}`)

    if (frontMatter.series_order !== undefined) {
      lines.push(`series_order: ${frontMatter.series_order}`)
    }
  }

  return lines.join('\n')
}

/**
 * 提取 Markdown 内容（不含 Front Matter）
 * @param content Markdown 文件完整内容
 * @returns 正文内容
 */
export function extractContent(content: string): string {
  const { content: body } = parseMatter(content)
  return body
}

/**
 * 组合 Front Matter 和 Markdown 内容
 * @param frontMatter Front Matter 对象
 * @param body 正文内容
 * @returns 完整的 Markdown 文件内容
 */
export function buildMarkdown(frontMatter: Partial<FrontMatter>, body: string = ''): string {
  const yamlStr = serializeFrontMatter(frontMatter)
  return `---\n${yamlStr}\n---\n${body}`
}

/**
 * 转义 YAML 字符串
 * 处理包含特殊字符的字符串
 */
function escapeYamlString(str: string): string {
  if (!str) return str

  // 如果字符串包含特殊字符，用引号包裹
  const specialChars = /[:#\n\r\t{}[\],&*?|<>=!%@`]/
  if (specialChars.test(str) || str.startsWith(' ') || str.endsWith(' ')) {
    // 转义双引号
    const escaped = str.replace(/"/g, '\\"')
    return `"${escaped}"`
  }

  return str
}

/**
 * 验证 Front Matter 是否有效
 * @param frontMatter Front Matter 对象
 * @returns 验证结果
 */
export function validateFrontMatter(frontMatter: Partial<FrontMatter>): {
  valid: boolean
  errors: string[]
} {
  const errors: string[] = []

  // 标题是必须的
  if (!frontMatter.title || frontMatter.title.trim() === '') {
    errors.push('标题不能为空')
  }

  // 日期格式验证
  if (frontMatter.date) {
    const dateRegex = /^\d{4}-\d{2}-\d{2}(\s+\d{2}:\d{2}:\d{2})?$/
    if (!dateRegex.test(frontMatter.date)) {
      errors.push('日期格式无效，应为 YYYY-MM-DD 或 YYYY-MM-DD HH:mm:ss')
    }
  }

  // 更新日期格式验证
  if (frontMatter.updated) {
    const dateRegex = /^\d{4}-\d{2}-\d{2}(\s+\d{2}:\d{2}:\d{2})?$/
    if (!dateRegex.test(frontMatter.updated)) {
      errors.push('更新日期格式无效，应为 YYYY-MM-DD 或 YYYY-MM-DD HH:mm:ss')
    }
  }

  // 状态验证
  if (frontMatter.status && !['draft', 'wip', 'published'].includes(frontMatter.status)) {
    errors.push('状态值无效，应为 draft、wip 或 published')
  }

  return {
    valid: errors.length === 0,
    errors
  }
}

/**
 * 更新 Front Matter
 * 保留原有内容，只更新指定字段
 * @param content 原始 Markdown 内容
 * @param updates 要更新的字段
 * @returns 更新后的 Markdown 内容
 */
export function updateFrontMatter(
  content: string,
  updates: Partial<FrontMatter>
): string {
  const { frontMatter, body } = parseFrontMatter(content)
  const updated = { ...frontMatter, ...updates }

  // 如果有更新，自动设置 updated 时间
  if (Object.keys(updates).length > 0 && !updates.updated) {
    updated.updated = new Date().toISOString().replace('T', ' ').slice(0, 19)
  }

  return buildMarkdown(updated, body)
}

/**
 * 获取默认 Front Matter
 * @param title 文章标题
 * @param categories 分类
 * @returns 默认 Front Matter 对象
 */
export function getDefaultFrontMatter(
  title: string = '未命名文章',
  categories: string[] = ['未分类']
): FrontMatter {
  const now = new Date()
  const dateStr = now.toISOString().replace('T', ' ').slice(0, 19)

  return {
    title,
    date: dateStr,
    categories,
    tags: [],
    description: '',
    status: 'draft',
    lang: 'zh-CN',
    comments: true,
    published: true
  }
}